src/fann_train_data.c

```diff
@@ -1186,8 +1186,10 @@ FANN_EXTERNAL void FANN_API fann_descale_train( struct fann *ann, struct fann_tr
 					- ann->scale_mean_##where[ cur_neuron ] 											\
 				); 																						\
 	for( cur_neuron = 0; cur_neuron < ann->num_##where##put; cur_neuron++ )								\
+		/* If all inputs are the same, set deviation to 1 instead of 0 to avoid division-by-zero */     \
 		ann->scale_deviation_##where[ cur_neuron ] =													\
-			sqrtf( ann->scale_deviation_##where[ cur_neuron ] / (float)data->num_data ); 			\
+			ann->scale_deviation_##where[ cur_neuron ] == 0.0f ? 1.0 :                                  \
+				sqrtf( ann->scale_deviation_##where[ cur_neuron ] / (float)data->num_data ); 			\
 	/* Calculate factor: (new_max-new_min)/(old_max(1)-old_min(-1)) */									\
 	/* Looks like we dont need whole array of factors? */												\
```

Comment:
Isn't it better to set it to 0.0f instead of 1.0, since the developer expects the input to be 0.0f?

I'm not entirely certain if this is the only place you get the divided by zero issue, but if it is, then only dividing in the "else block", as youve done, should be enough?

Reply:
The division by `data->num_data` here is actually unproblematic, the functions `fann_set_input_scaling_params()` and `fann_set_output_scaling_params()` already guard against this case. (and set mean and deviation to the default values 0.0 and 1.0 resp.)

The problem described in issue #99 happens in `fann_scale_train()` and `fann_descale_train()`. The former divides by the standard deviation, the latter mulitplies by it. This can happen if the values of one input or output column are all the same and we currently don't guard against that. Good catch though, I'll expand the comment above this line to point to the problematic functions.

Reply:
Not sure if it's a good idea, but to make it more obvious what's going on, we could also put the check into a separate statement. Something like this:

```c
for( cur_neuron = 0; cur_neuron < ann->num_##where##put; cur_neuron++ ) {
  ann->scale_deviation_##where[ cur_neuron ] =
    sqrtf( ann->scale_deviation_##where[ cur_neuron ] / (float)data->num_data );
  if( ann->scale_deviation_##where[ cur_neuron ] == 0.0f ) {
    /* explanatory comment here */
    ann->scale_deviation_##where[ cur_neuron ] = 1.0f;
  }
}
```
But using the ternary operator as in the original commit seemed more ~efficient~ to me.

Reply:
when reviewing the issue I see that he ignored every neuron with a val of 0.0f. As they caused an issue:
```
if(ann->scale_deviation_in[ cur_neuron ] != 0.0)
```

But now those are changed to 1.0f later. My concern is what happens to later code, that has been developed with it being 0.0f in mind?

Reply:
My approach differs from the one suggested in issue #99. Their solution was: if the standard deviation of an input is zero, don't scale it at all. My solution still scales such an input, but with an (admittedly arbitrarily chosen) standard deviation of 1.0.

The advantage of the original solution is that it's certainly less surprising. The advantage of mine is that even in the case of the input being constant, it is scaled into the target range `[new_min; new_max]`.

As far as I can tell, we're pretty much free to choose either option, since the docs explicitly don't guarantee that the input will be in-range after scaling. I'm not having a strong opinion either and the change to the solution originally suggested isn't too complicated.

Would you prefer we solve this case as originally suggested by not scaling these degenerate inputs at all?

Reply:
I'm leaning towards to ignore cases with 0.0f:
```
ann->scale_deviation_##where[ cur_neuron ] == 0.0f ? 0.0f : sqrtf( ann->scale_deviation_##where[ cur_neuron ] / (float)data->num_data );
```

Reply:
Thoughts? @Bukka

Reply:
Would it be sensible to disable an input neuron with 0 deviation ?

Reply:
I would be more in favor of keeping that as `0.0f` and ignore those neurons in scaling. The thing is that if you already  saved such `ann` by `fann_save` (let's say before you updated the library with the new fix changing that to 1.0f) and then loaded `ann` by `fann_load` after you did the lib update, it will still have value 0.0f so we would still need to add those checks to `fann_descale` and `fann_scale`...

Also It seems slightly safer in case some code out there would be using `scale_deviation_[in|out]` explicitly and relied on it to be a zero but that's quite unlikely IMHO.